metrics/engine/engine.go

```diff
@@ -0,0 +1,174 @@
+// Package engine contains the internal metrics engine responsible for
+// aggregating metrics during the test and evaluating thresholds against them.
+package engine
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+	"go.k6.io/k6/lib"
+	"go.k6.io/k6/metrics"
+	"go.k6.io/k6/output"
+	"go.k6.io/k6/stats"
+	"gopkg.in/guregu/null.v3"
+)
+
+// MetricsEngine is the internal metrics engine that k6 uses to keep track of
+// aggregated metric sample values. They are used to generate the end-of-test
+// summary and to evaluate the test thresholds.
+type MetricsEngine struct {
+	registry       *metrics.Registry
+	executionState *lib.ExecutionState
+	options        lib.Options
+	runtimeOptions lib.RuntimeOptions
+	logger         logrus.FieldLogger
+
+	// These can be both top-level metrics or sub-metrics
+	metricsWithThresholds []*stats.Metric
+
+	// TODO: completely refactor:
+	//   - make these private,
+	//   - do not use an unnecessary map for the observed metrics
+	//   - have one lock per metric instead of a a global one, when
+	//     the metrics are decoupled from their types
+	MetricsLock     sync.Mutex
+	ObservedMetrics map[string]*stats.Metric
+}
+
+// NewMetricsEngine creates a new metrics Engine with the given parameters.
+func NewMetricsEngine(
+	registry *metrics.Registry, executionState *lib.ExecutionState,
+	opts lib.Options, rtOpts lib.RuntimeOptions, logger logrus.FieldLogger,
+) (*MetricsEngine, error) {
+	me := &MetricsEngine{
+		registry:       registry,
+		executionState: executionState,
+		options:        opts,
+		runtimeOptions: rtOpts,
+		logger:         logger.WithField("component", "metrics-engine"),
+
+		ObservedMetrics: make(map[string]*stats.Metric),
+	}
+
+	if !(me.runtimeOptions.NoSummary.Bool && me.runtimeOptions.NoThresholds.Bool) {
+		err := me.initSubMetricsAndThresholds()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return me, nil
+}
+
+// GetIngester returns a pseudo-Output that uses the given metric samples to
+// update the engine's inner state.
+func (me *MetricsEngine) GetIngester() output.Output {
+	return &outputIngester{
+		logger:        me.logger.WithField("component", "metrics-engine-ingester"),
+		metricsEngine: me,
+	}
+}
+
+func (me *MetricsEngine) getOrInitPotentialSubmetric(name string) (*stats.Metric, error) {
+	// TODO: replace with strings.Cut after Go 1.18
+	nameParts := strings.SplitN(name, "{", 2)
```

Comment:
I ended up doing something somewhat similar to extract metric/submetric name for thresholds validation. I ended up using `strings.FieldFunc`, which, I believe, could lead to somewhat clearer code here?

```go
delimiterFn := func(c rune) bool {
    return c == '{' || c == '}' || c == ':'
}

fmt.Println("Fields are: %q", strings.FieldsFunc("metric_name{foo:bar}", f))
```

The resulting length is either 1 (metric), or an odd value >=3. No need to match the closing `}` for instance. You could also drop the matching of `:` in your case, I believe.

Hope that's helpful 🙇🏻

Reply:
Wouldn't that also accept `metric_name}foo{bar:` as valid? :confused:

Reply:
That's a fair point. I need to look into it. I'll add a test for it in https://github.com/grafana/k6/pull/2463. Commit https://github.com/grafana/k6/pull/2463/commits/e4e51f0a4207154ba3d42137cc283ddd9c382222 introduces a `ParseMetricName` function.